freedomofpress/securedrop
# Supports optional HTTPS on Source Interface

## Status

Ready for review.

## Description of Changes

Fixes #1605. Several organizations have started deploying HTTPS on the Source Interface, and we've worked with them to create custom vhost configs. Given that the config has been stable in production, we can now add top-level configuration to enable HTTPS if orgs wish to opt-in.

Most of the configuration must happen out-of-band, via manual changes on the Application Server. Since DigiCert provides only EV certificates for Onion URLs, the Admin will need to communicate with DigiCert directly and confirm ownership of the URL. The documentation provided in this PR links out to the relevant DigiCert contact pages, and explicitly mentions the Ansible vars that should be updated, but is light on details concerning the nitty-gritty of the EV process. I'm fine with that for now.

The config tests have been updated, as well. Noticed that in the Apache test suite, we were still hardcoding a few vars, rather than importing dynamically from the YAML vars file. Fixed that, and cleaned up some of the Apache templates while in there (056e96574630147a043fee65030b9d290c316c72). 

## Testing

1. Provision staging hosts as usual, and confirm you can access the Source Interface over HTTP at the Onion URL.
2. Generate a self-signed keypair (e.g. with `easy-rsa`) for use with the Source Interface. You should also generate a CA crt, for a total of three files that will land on the Application Server.
3. Update the site-specific vars as described in the new HTTPS docs. (I'm being intentionally vague here; if the docs aren't clear, request changes in the review.)
4. Rerun the config: `ANSIBLE_ARGS="--tags app,tor" vagrant provision /staging/`.
5. Confirm you can access the Source Interface over HTTPS. (You'll need to accept the self-signed cert.)
6. Confirm the HTTP URL redirects automatically to HTTPS.

## Deployment

For the few orgs already using HTTPS, we should reach out and describe the new vars flow during the 0.4 roll-out. For new orgs, the opt-in flow will default to the normal HTTP configuration, so no changes unless Admins wish to obtain an EV cert.

## Checklist

**Relying on manual testing to confirm functionality.** We don't have config testing for self-signed certs right now. In general, we're not testing multiple configuration paths in the Ansible setup, although that's something we should definitely start doing in the future.

### If you made changes to the app code:


### If you made changes to the system configuration:


### If you made changes to documentation:
